Dockerfile

```diff
@@ -33,28 +46,24 @@ RUN mkdir -p /usr/src/app && \
 
 WORKDIR /usr/src/app
 
-COPY Gemfile* .ruby-version ./
+# copy over gems from the dependencies stage
+COPY --from=dependencies /usr/local/bundle/ /usr/local/bundle/
 
-RUN gem install bundler && \
-    bundle config set frozen 'true' && \
-    bundle config without test:development && \
-    bundle install --jobs 2 --retry 3
+# copy over npm packages from the dependencies stage
+COPY --from=dependencies /node_modules/ node_modules/
 
+# copy over the remaning files and code
 COPY . .
 
-# The following are ENV variables that need to be present by the time
+# Some ENV variables need to be present by the time
 # the assets pipeline run, but doesn't matter their value.
-ENV SECRET_KEY_BASE=needed_for_assets_precompile
-ENV ENV_NAME=needed_for_assets_precompile
+RUN SECRET_KEY_BASE=needed_for_assets_precompile \
+    RAILS_ENV=production \
```

Comment:
Always seems an unsatisfactory solution to me, but I've never found a better one. 

Initially I was unhappy with how you've used production for the stub.  But thinking about it, I wonder if using production, as you have done, may be safest as at least it would prevent a test run against the db etc. Was that your reasoning?

Reply:
These values here doesn't really matter. However, they have to be present for `rails assets:precompile` to work. And they need to "make some sense" (so to speak) as the command will invoke a rails environment. So for instance, `RAILS_ENV` will be checked against `config/environments/*.rb` (development, test and production). So `RAILS_ENV=foobar` will blow up.

And we want to ensure assets are compiled for a production target. Because the compilation could tweak things like _minifying_, _uglifying_, etc. depending on the environment.

The real ENV values for the actual running of the image are always taken from the Kubernetes environment (coming either from config_map, or secrets). Or from the `docker-compose.yml` when running the container locally (`docker-compose up`). Docker compose is never used in the cluster, is only there for our convenience.

Hope it makes sense? Let me know if it is easier to have a chat about it?